kotlin-codegen/compiler/src/main/java/com/squareup/moshi/MoshiKotlinCodeGenProcessor.kt

```diff
@@ -0,0 +1,759 @@
+            .addStatement("%N.nextName()", reader)
+            .addStatement("%N.skipValue()", reader)
+            .endControlFlow()
+            .endControlFlow()
+            .endControlFlow()
+            .addStatement("%N.endObject()", reader)
+            .apply {
+              val propertiesWithDefaults = localProperties.entries.filter { it.key.hasDefault }
+              val propertiesWithoutDefaults = localProperties.entries.filter { !it.key.hasDefault }
+              val requiredPropertiesCodeBlock = CodeBlock.of(
+                  propertiesWithoutDefaults.joinToString(",\n") { (property, specs) ->
+                    val spec = specs.first
+                    "${property.name} = ${spec.name}%L"
+                  },
+                  *(propertiesWithoutDefaults
+                      .map { (property, _) ->
+                        if (property.isRequired) {
+                          @Suppress("IMPLICIT_CAST_TO_ANY")
+                          CodeBlock.of(
+                              " ?: throw %T(\"Required property '%L' missing at \${%N.path}\")",
+                              JsonDataException::class.asTypeName(),
+                              property.name,
+                              reader
+                          )
+                        } else {
+                          @Suppress("IMPLICIT_CAST_TO_ANY")
+                          ""
+                        }
+                      }
+                      .toTypedArray()))
+              if (propertiesWithDefaults.isEmpty()) {
+                addStatement("return %T(%L)",
+                    originalTypeName,
+                    requiredPropertiesCodeBlock)
+              } else {
+                addStatement("return %T(%L)\n.let {\n  it.copy(%L)\n}",
+                    originalTypeName,
+                    requiredPropertiesCodeBlock,
+                    propertiesWithDefaults
+                        .joinToString(",\n      ") { (property, specs) ->
+                          val spec = specs.first
+                          val setSpec = specs.second
+                          if (setSpec != null) {
+                            "${property.name} = if (${setSpec.name}) ${spec.name} else it.${property.name}"
+                          } else {
+                            "${property.name} = ${spec.name} ?: it.${property.name}"
+                          }
+                        })
+              }
+            }
+            .build())
+        .addFunction(FunSpec.builder("toJson")
+            .addModifiers(OVERRIDE)
+            .addParameter(writer)
+            .addParameter(value)
+            .addStatement("%N.beginObject()", writer)
+            .apply {
+              propertyList.forEachIndexed { index, prop ->
+                addStatement("%N.name(%S)",
+                    writer,
+                    prop.serializedName)
+                addStatement("%N.toJson(%N, %N${if (index == 0) "!!" else ""}.%L)",
```

Comment:
No need to emit `!!`. Just mark the parameter as non-null.

Reply:
@swankjesse and I talked about this a bit. That might work in gradle, but I don't see how to do that in a maven project. It doesn't see the JSR305 bit, so it fails to compile if you make the parameter nonnull

Reply:
The only way that it would fail to compile is if it did see a `@Nullable` annotation directly on the parameter.

Reply:
There is a `@Nullable` annotation on the parameter.

I would just emit a manual null check as the first statement then rather than complicate the implementation with this logic.

Reply:
Fair enough. Went with a more explicit nullcheck but let me know if you'd rather it just be a `value!!` at the beginning. 5ddf528

Reply:
Nope. Like what you got.

Reply:
:+1: